File: python/diamond.py
def diff(c1, c2):
    return ord(c1) - ord(c2)

def row(ch, outer_len):
    outer = ' ' * outer_len
    if ch == 'A':
        return f'{outer}{ch}{outer}'

    inner_len = diff(ch, 'A')
    inner = ' ' * (inner_len * 2 - 1)
    return f'{outer}{ch}{inner}{ch}{outer}'
    
def rows(letter):
    if letter == 'A':
        return ['A']

    res = []

    ch = 'A'
    outer_len = diff(letter, 'A')
    
    while ch < letter:
        res.append(row(ch, outer_len))
        ch = chr(ord(ch) + 1)
        outer_len -= 1
    
    while ch >= 'A':
        res.append(row(ch, outer_len))
        ch = chr(ord(ch) - 1)
        outer_len += 1
        
    return res

File: dart/diamond.dart
class Diamond {
  List<String> rows(String letter) {
    List<String> res = [];

    String ch = 'A';
    int outerLen = diff(letter, ch);

    while (ch.compareTo(letter) < 0) {
      res.add(row(ch, outerLen));
      ch = String.fromCharCode(ch.codeUnitAt(0) + 1);
      outerLen--;
    }

    while (ch.compareTo('A') >= 0) {
      res.add(row(ch, outerLen));
      ch = String.fromCharCode(ch.codeUnitAt(0) - 1);
      outerLen++;
    }
    return res;
  }

  int diff(String a, String b) {
    return a.codeUnitAt(0) - b.codeUnitAt(0);
  }

  String row(String ch, int outerLen) {
    String outer = ' ' * outerLen;
    if (ch == 'A') {
      return '$outer$ch$outer';
    }
    int innerLen = diff(ch, 'A') * 2 - 1;
    String inner = ' ' * innerLen;
    return '$outer$ch$inner$ch$outer';
  }
}
